### PR TITLE
[SPARK-18624][SQL] Implicit cast ArrayType(InternalType)

### PR DESCRIPTION
**Pull request by jiangxb1987**
## What changes were proposed in this pull request?

Currently `ImplicitTypeCasts` doesn't handle casts between `ArrayType`s, this is not convenient, we should add a rule to enable casting from `ArrayType(InternalType)` to `ArrayType(newInternalType)`.

Goals:
1. Add a rule to `ImplicitTypeCasts` to enable casting between `ArrayType`s;
2. Simplify `Percentile` and `ApproximatePercentile`.

## How was this patch tested?

Updated test cases in `TypeCoercionSuite`.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 1]
```diff
@@ -714,6 +714,17 @@ object TypeCoercion {
         // try to find the first one we can implicitly cast.
         case (_, TypeCollection(types)) => types.flatMap(implicitCast(e, _)).headOption.orNull
 
+        case (ArrayType(_, nullable), ArrayType(internalType: DataType, expectedNullable))
+          if (expectedNullable || nullable == expectedNullable) =>
+          val newChildren = e.children.map { expr =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Why is this needed? We can cast the array directly.

**Review by hvanhovell — CHANGES_REQUESTED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 2]
```diff
@@ -714,6 +714,17 @@ object TypeCoercion {
         // try to find the first one we can implicitly cast.
         case (_, TypeCollection(types)) => types.flatMap(implicitCast(e, _)).headOption.orNull
 
+        case (ArrayType(_, nullable), ArrayType(internalType: DataType, expectedNullable))
+          if (expectedNullable || nullable == expectedNullable) =>
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Ah, actually @hvanhovell, in case of `case _ if expectedType.acceptsType(inType) => inType` would this return `None` instead of the same input type? 

I see this is being used by `implicitCast(in, expected).getOrElse(in)` so it would be fine to work but I am worried of this (e.g. for unit test and etc).

**Comment by viirya** [comment]
If this is going to support ArrayType only, can you update the pr title to reflect it?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala (hunk 2 above)
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: We should also consider the case forcing the casting result is null. Please see https://github.com/apache/spark/blob/7eb2ca8e338e04034a662920261e028f56b07395/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala#L94

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: @HyukjinKwon the idea is that returning `None` means that we should not cast at all. Returning the same type would mean that we could cast. This is only relevant for `ArrayType` case, because we could only change the nullability of the column, e.g.: `Cast(a, ArrayType(DoubleType, false), ArrayType(DoubleType, true))`

Replacing the `collect { ... }` by a `map { ... }` changes the semantic of the `implicitCast(...)` function, and will probably break the [`ResolveTableValuedFunctions`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveTableValuedFunctions.scala#L39-L49) rule.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: @viirya yeah, we should consider the nullability.

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Thank you for your detailed explanation and yeap that makes sense. I see it is up to what `None` means from the function. I was just worried of changing the existing behaviour and wanted to note just in case. I agree with either way if it sounds consistent:).

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: I just ran `TypeCoercionSuite` with the change and it seems several tests are failed when the source and target types are the same.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: @HyukjinKwon it is a suggestion, I did not test it. I am curious to see what is failing.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Running it now....:)

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: Thank you!!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala [hunk 3]
```diff
@@ -673,48 +673,55 @@ object TypeCoercion {
...
+        case (any: AtomicType, StringType) if any != StringType => StringType
 
         // When we reach here, input type is not acceptable for any types in this type collection,
         // try to find the first one we can implicitly cast.
-        case (_, TypeCollection(types)) => types.flatMap(implicitCast(e, _)).headOption.orNull
+        case (_, TypeCollection(types)) =>
+          types.flatMap(implicitCast(inType, _)).headOption.orNull
+
+        case (ArrayType(fromType, fn), ArrayType(toType: DataType, tn))
+            if !(fn || Cast.forceNullable(fromType, toType)) || tn =>
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala]: I had to look quite long at this one, so:

- `toNull` => Always allow
- `fromNull && !toNull` => Never allow
- `!fromNull && !toNull` => Only allow when `!Cast.forceNullable(fromType, toType)`

What you have written is just a very terse representation of that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala [hunk 4]
```diff
@@ -85,23 +85,16 @@ case class ApproximatePercentile(
   private lazy val accuracy: Int = accuracyExpression.eval().asInstanceOf[Int]
 
   override def inputTypes: Seq[AbstractDataType] = {
-    Seq(DoubleType, TypeCollection(DoubleType, ArrayType), IntegerType)
+    Seq(DoubleType, TypeCollection(DoubleType, ArrayType(DoubleType)), IntegerType)
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/ApproximatePercentile.scala]: `ArrayType(DoubleType, false)`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala [hunk 5]
```diff
@@ -98,7 +92,7 @@ case class Percentile(
   }
 
   override def inputTypes: Seq[AbstractDataType] = percentageExpression.dataType match {
-    case _: ArrayType => Seq(NumericType, ArrayType)
+    case _: ArrayType => Seq(NumericType, ArrayType(DoubleType))
```
Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala]: `ArrayType(DoubleType, false)`?

Comment (jiangxb1987) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala]: If the input array's widest type is a `DecimalType`, `CreateArray()` will create a `ArrayType(DecimalType, containsNull = true)`, because the function `Cast. forceNullable(fromType, toType)` always returns true if `toType` is a `DecimalType`.
Then, if we are expecting a inputType as `ArrayType(DoubleType, containsNull = false)`, we fail to cast because the nullability is not compatible.
For example, the `array(0.0, 1.0)` will be casted to `ArrayType(DecimalType(2, 1), containsNull = true)`, and this is a valid input for `Percentile` function.
Thus, I decide to keep the `inputType` to be `ArrayType(DoubleType, containsNull = true)`, to be compatible with the above case.

Comment (hvanhovell) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/Percentile.scala]: Ok, that is reasonable.

**Comment by hvanhovell** [comment]
LGTM - merging to master. Thanks!

**Comment by ashokblend** [comment]
Hi Guys

**Comment by ashokblend** [comment]
Any Reason why its not merged in branch2.1
